bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieWriteToJournalTest.java

```diff
@@ -0,0 +1,112 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.bookie;
+
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.powermock.api.support.membermodification.MemberMatcher.method;
+import static org.powermock.api.support.membermodification.MemberModifier.replace;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import java.io.File;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.conf.TestBKConfiguration;
+import org.apache.bookkeeper.net.BookieSocketAddress;
+import org.apache.bookkeeper.proto.BookkeeperInternalCallbacks.WriteCallback;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+/**
+ * Test the bookie journal.
+ */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({Journal.class})
+@Slf4j
+public class BookieWriteToJournalTest {
+
+    private static final ByteBuf DATA = Unpooled.buffer();
+    static {
+        DATA.writeLong(1); // ledgerId
+        DATA.writeLong(1); // entryId
+    }
+
+    @Rule
+    public TemporaryFolder tempDir = new TemporaryFolder();
+
+    /**
+     * test that Bookie calls correctly Journal.logAddEntry about "ackBeforeSync" parameter.
+     */
+    @Test
+    public void testJournalLogAddEntryCalledCorrectly() throws Exception {
+        File journalDir = tempDir.newFolder();
+        Bookie.checkDirectoryStructure(Bookie.getCurrentDirectory(journalDir));
+        File ledgerDir = tempDir.newFolder();
+        Bookie.checkDirectoryStructure(Bookie.getCurrentDirectory(ledgerDir));
+        ServerConfiguration conf = TestBKConfiguration.newServerConfiguration();
+        conf.setJournalDirName(journalDir.getPath())
+            .setLedgerDirNames(new String[]{ledgerDir.getPath()})
+            .setZkServers(null);
+        Bookie b = new Bookie(conf);
+        b.start();
+        Boolean[] effectiveAckBeforeSync = new Boolean[1];
+        replace(method(Journal.class, "logAddEntry", ByteBuf.class, Boolean.TYPE, WriteCallback.class, Object.class))
```

Comment:
why not use PowerMock to mock journal constructor? so you will have mocked journal object. so that you can assert journal method calls with expected flags.

Reply:
@sijie  @ivankelly @yzang 

I have updated this test and now I am using a full mock Journal, now the test is super "flaky" I think due to the caching of FileInfo, if I run the test with a debugger it passes.

I wonder if I have to spend time in fixing the flakiness or revert to previous test or wait for fixes on FileInfo cache flakyness

```
org.apache.bookkeeper.bookie.Bookie$NoLedgerException: Ledger 1 not found
	at org.apache.bookkeeper.bookie.IndexPersistenceMgr$FileInfoLoader.call(IndexPersistenceMgr.java:226)
	at org.apache.bookkeeper.bookie.IndexPersistenceMgr$FileInfoLoader.call(IndexPersistenceMgr.java:211)
	at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4904)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3628)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2336)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2295)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2208)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4053)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4899)
	at org.apache.bookkeeper.bookie.IndexPersistenceMgr.getFileInfo(IndexPersistenceMgr.java:257)
	at org.apache.bookkeeper.bookie.IndexPersistenceMgr.isFenced(IndexPersistenceMgr.java:499)
	at org.apache.bookkeeper.bookie.LedgerCacheImpl.isFenced(LedgerCacheImpl.java:147)
	at org.apache.bookkeeper.bookie.InterleavedLedgerStorage.isFenced(InterleavedLedgerStorage.java:208)
	at org.apache.bookkeeper.bookie.LedgerDescriptorImpl.isFenced(LedgerDescriptorImpl.java:76)
	at org.apache.bookkeeper.bookie.Bookie.addEntry(Bookie.java:1361)
	at org.apache.bookkeeper.bookie.BookieWriteToJournalTest.testJournalLogAddEntryCalledCorrectly(BookieWriteToJournalTest.java:110)
```

Reply:
@ivankelly found the problem, the Bookie exists with the mock journal, I have to mock the Journal#join method